snuba/query/joins/pre_processor.py

```diff
@@ -0,0 +1,166 @@
+            if relationship.rhs_entity in self.__entities:
+                for equivalence in relationship.equivalences:
+                    self.__add_relationship(
+                        ret,
+                        _entity_from_node(node),
+                        equivalence.left_col,
+                        relationship.rhs_entity,
+                        equivalence.right_col,
+                    )
+        return ret
+
+    def visit_join_clause(self, node: JoinClause[Entity]) -> EquivalencesGraph:
+        ret: EquivalencesGraph = {}
+        mapping = node.get_alias_node_map()
+        for condition in node.keys:
+            self.__add_relationship(
+                ret,
+                _entity_from_node(mapping[condition.left.table_alias]),
+                condition.left.column,
+                _entity_from_node(mapping[condition.right.table_alias]),
+                condition.right.column,
+            )
+
+        def merge_into_graph(node: JoinNode[Entity]) -> None:
+            for col, equivalences in node.accept(self).items():
+                if col not in ret:
+                    ret[col] = equivalences
+                else:
+                    ret[col] |= equivalences
+
+        merge_into_graph(node.left_node)
+        merge_into_graph(node.right_node)
+        return ret
+
+
+def get_equivalent_columns(
+    join: JoinClause[Entity],
+) -> Mapping[QualifiedCol, Set[QualifiedCol]]:
+    """
+    Given a Join, it returns the set of all the semantically equivalent
+    columns across the entities involved in the join.
+
+    This is obtained by generating, through EquivalenceExtractor, the
+    graph of all equivalences.
+    We then have the sets of semantically equivalent columns by
+    generating the list of connected components in the equivalence graph
+
+    Each node in a connected component of the equivalence graph is by
+    definition, semantically equivalent to all the nodes of the same
+    connected component (directly if there is an edge between two columns
+    or transitively).
+
+    The connected components are returned as a Mapping of nodes to their
+    connected component (which is a set of nodes).
+    """
+
+    def traverse_graph(
+        node: QualifiedCol, visited_nodes: Set[QualifiedCol]
+    ) -> Set[QualifiedCol]:
+        """
+        Traverse the whole connected component in with a depth first
+        algorithm starting from the node provided.
+        """
+        if node in visited_nodes:
+            return visited_nodes
+        visited_nodes.add(node)
+        for n in adjacency_sets.get(node, set()):
+            visited_nodes = traverse_graph(n, visited_nodes)
+        return visited_nodes
+
+    entities_in_join = {
+        _entity_from_node(node) for node in join.get_alias_node_map().values()
+    }
+    adjacency_sets = join.accept(EquivalenceExtractor(entities_in_join))
+    connected_components: MutableMapping[QualifiedCol, Set[QualifiedCol]] = {}
```

Comment:
Is there a reason you didn't want to use the EquivalenceGraph type for this?

Reply:
The only reason was naming. This is a map between a node and its connected component (expressed as a set of nodes) not a graph. EquivalenceGraph is an adjacency set representation of a graph, the sets here do not represent an adjacency set. The return value just happen to have syntactically the same structure as EquivalenceGraph but it is not semantically a graph.